toolkit/tools/pkg/licensecheck/licensecheck.go

```diff
@@ -0,0 +1,411 @@
+  - Call GetAllResults() to get all the results, split into buckets.
+
+4. Call CleanUp() to tear down the chroot
+
+Also provided are several directly callable functions (these expect to be run in an environment with the necessary
+macros, i.e. a chroot): CheckRpmLicenses(), IsALicenseFile(), IsASkippedLicenseFile()
+
+The LicenseCheckerResult struct is used to store the results of the search. It contains the path to the RPM, a list of
+bad documents, a list of bad files, and a list of duplicated documents. The bad documents are %doc files that are not
+at least also in the license files. The bad files are general files that are misplaced in the licenses directory.
+
+The duplicated documents are %doc files that are also in the license files. These are not technically bad, but are messy
+and should be cleaned up.
+*/
+package licensecheck
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/logger"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/rpm"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/sliceutils"
+	"github.com/microsoft/azurelinux/toolkit/tools/pkg/simpletoolchroot"
+)
+
+const (
+	chrootName      = "license_chroot"
+	licensePrefix   = "/usr/share/licenses"
+	numParallelJobs = 20
+)
+
+// LicenseChecker is a tool for searching RPMs for bad licenses
+type LicenseChecker struct {
+	simpleToolChroot *simpletoolchroot.SimpleToolChroot // The chroot to scan the RPMs in
+	distTag          string                             // The distribution tag to use when parsing RPMs
+	licenseNames     LicenseNames                       // The regexes used to match license files
+	exceptions       LicenseExceptions                  // Files that should be ignored
+	results          []LicenseCheckResult               // The results of the search
+	jobSemaphore     chan struct{}                      // Limit the number of parallel jobs
+}
+
+// New creates a new license checker. If this returns successfully the caller is responsible for calling CleanUp().
+// - buildDirPath: The path to create the chroot inside
+// - workerTarPath: The path to the worker tarball
+// - rpmDirPath: The path to the directory containing the RPMs
+// - nameFilePath: The path to the .json file containing license names
+// - exceptionFilePath: Optional, the path to the .json file containing license exceptions to ignore
+// - distTag: The distribution tag to use when parsing RPMs
+func New(buildDirPath, workerTarPath, rpmDirPath, nameFilePath, exceptionFilePath, distTag string,
+) (newLicenseChecker *LicenseChecker, err error) {
+	newLicenseChecker = &LicenseChecker{
+		distTag:          distTag,
+		simpleToolChroot: &simpletoolchroot.SimpleToolChroot{},
+		jobSemaphore:     make(chan struct{}, numParallelJobs),
+	}
+
+	err = newLicenseChecker.simpleToolChroot.InitializeChroot(buildDirPath, chrootName, workerTarPath, rpmDirPath)
+	if err != nil {
+		newLicenseChecker.CleanUp()
```

Comment:
The fact that we have to run `newLicenseChecker.CleanUp()` when `InitializeChroot` fails seems like an issue in `InitializeChroot`, which should clean-up after itself, if it fails.

Reply:
Probably just being overly defensive. One of those spots where it feels like we should err on the side of caution and always clean up after ourselves even if in this specific case it may not be needed today.